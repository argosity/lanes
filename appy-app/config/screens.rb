
Lanes::Screen.define "base" do | screen |
    screen.description = "Base Screen for Appy App"
    screen.icon        = ""
    screen.group_id    = ""
    screen.model_class = ""
    screen.view_class  = "AppyApp.Screens.Base"
end
