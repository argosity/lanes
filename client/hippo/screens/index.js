import { observable, computed } from 'mobx';
import { map, filter, intersection } from 'lodash';
import Group from './group';
import Config from '../config';

const Screens = observable({

    all: observable.map(),

    get activeGroups() {
        const gids = Group.enabled_group_ids || map(Group.all, 'id');
        const groups = map(gids, gid => Group.forId(gid));
        return filter(groups, group => intersection(group.screens, this.active).length);
    },

    @computed get active() {
        return map(Config.screen_ids, id => this.all.get(id));
    },

    reset() {
        Config.screen_ids.clear();
    },

});

export default Screens;
