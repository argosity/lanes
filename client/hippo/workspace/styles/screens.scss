.screen {
    padding: 5px;
    width: 100%;

    > .first-run {
        font-size: 1.3rem;
        li + li {
            margin-top: 1rem;
        }
        .help {
            margin-top: 2rem;
            p { text-align: right; }
        }
    }
}
