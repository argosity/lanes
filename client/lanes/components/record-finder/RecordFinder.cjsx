class Lanes.Components.RecordFinder extends Lanes.React.Component

    propTypes:
        query:    Lanes.PropTypes.State.isRequired
        model:    Lanes.PropTypes.State.isRequired
        commands: React.PropTypes.object.isRequired


    contextTypes:
        viewport: Lanes.PropTypes.State.isRequired

    showFinder: ->
        body = Lanes.u.withReactContext @context, =>
            <LC.RecordFinder.Dialog query={@props.query} onRecordSelect={@props.commands.setModel} />

        @context.viewport.displayModal(
            body: body
            title: "Find #{@props.query.title}"
            buttons: [{title: 'Cancel'}]
        ).then(Lanes.emptyFn, Lanes.emptyFn)

    loadCurrentSelection: ->
        value = @props.model[@props.query.initialField.id]
        @props.query.loadSingle(value).then (model) =>
            @props.commands.setModel(model)

    onKeyPress: (ev) ->
        if "Enter" == ev.key
            ev.stopPropagation()
            this.loadCurrentSelection()
        null

    render: ->
        findIcon = <button className="btn btn-primary icon icon-search icon-lg" onClick={@showFinder}/>
        <LC.Input
            ref="input"
            {...@props}
            groupClassName="record-finder"
            editOnly writable
            name={@props.query.initialField.id}
            onKeyPress={@onKeyPress}
            model={@props.model}
            buttonAfter={findIcon} />
