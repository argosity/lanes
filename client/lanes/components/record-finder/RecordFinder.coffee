class Lanes.Components.RecordFinder extends Lanes.Components.Base

    writeTemplateName: 'field'
    writeTemplateData: ->
        { fieldName: @fieldName || @recordQuery.initialField.id }

    domEvents:
        "keyup .record-finder-query-string": "onKey"
        "click .record-finder-query": "displayFinder"

    session:
        withAssociations: 'array'
        invalid_chars: 'regex'
        recordQuery: 'model'
        gridOptions: 'gridOptions'
        fieldName:  'string'
        queryField: [ 'string', false, 'code' ]

    ui:
        input: '.record-finder-query-string'

    constructor: (options={})->
        super
        this.access = 'write'

    initialize:(options)->
        @recordQuery = new Lanes.Models.Query(
            fields: options.fields, modelClass: options.modelClass
        )

    displayFinder: ->
        finder = new Lanes.Components.RecordFinder.Dialog(
            gridOptions: @gridOptions, parent:this
            title: @title, recordQuery: @recordQuery
        )
        finder.show().then( (dlg)->
            dlg.remove().record
        ).then( (record)=>
            if record
                if _.isEmpty(@withAssociations)
                    record.fetch()
                else
                    record.withAssociations(@withAssociations...)
            else
                null
        ).then( (reply)=>
            @.$el.trigger("display-record", reply.record) if reply?.record
        , (e)->Lanes.warn(e) )

    onKey: (ev)->
        if 13 == ev.keyCode
            code = this.$(ev.target).val()
            this.loadCode(code)
        else if @invalid_chars && @ui.input.val().match( @invalid_chars )
            @ui.input.val( @ui.input.val().replace( @invalid_chars, '' ) )

    loadCode: (code)->
        @recordQuery.loadSingle(code.toUpperCase(),{ include: @withAssociations })
            .then( (reply)=>
                this.$el.trigger("display-record", reply.record) if reply.record
            )
