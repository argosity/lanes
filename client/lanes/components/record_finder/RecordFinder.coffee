class FinderClause
    constructor: -> super

    template: 'record-finder/clause'

    initialize: ->
        window.fc = this.model

    events:
        'hidden.bs.dropdown': 'onQueryChange'
        'click .del-clause': 'delClause'
        'input input.query-string': 'onInput'

    onInput: (ev)->
        this.model.value = ev.target.value

    bindings:
        'model.description': '.query-field-description'

    subviews:
        fields:
            component: 'RadioGroup'
            data: 'model.fields'
            container: '.fields'
            options: { mappings: { id:'field'}, access: 'write' }

        operators:
            component: 'RadioGroup'
            data: 'model.operators'
            container: '.operators'
            options: { mappings: { title:'name', visible: 'valid' }, access: 'write' }

    delClause: ->
        @model.remove()

    onQueryChange: ->
        this.$('input.query-string').focus()

    render: ->
        super
        @defer @focus, delay: 500
        this

    focus: ->
        this.$('input.query-string').focus()


Lanes.Component.Base.extend(FinderClause)


class FinderDialog
    constructor: ->
        super

    events:
        'click .add-clause': 'addClause'
        'click .run-query':  'runQuery'
        'select-row .skr-grid': 'onSelect'

    subviews:
        grid:
            hook: 'grid'
            component: 'Grid'
            options: ->
                { record_query: @record_query }
        query_clauses:
             container: '.query-clauses'
             view: FinderClause, collection: 'clauses'

    bodyTemplateName: 'record-finder/dialog'
    bodyAttributes:
        class: "record-finder"

    session:
        record_query: 'model'
        clauses: 'collection'

    initialize:(options)->
        # @record_query = options.record_query
        # @columns     = options.columns
        # @model_type  = options.model_type

        @clauses = @record_query.clauses

        @debounceMethod( 'runQuery')
        this.listenTo(@record_query,'change',@runQuery)
        _.bindAll(this,'hide')

    onSelect: (ev)->
        @record = ev.detail.model
        ev.preventDefault()
        _.delay(this.hide, 500)

    addClause: ->
        @record_query.addNewClause()

    runQuery: ->
        this.grid.reload() if @record_query.isValid()

Lanes.Component.ModalDialog.extend(FinderDialog)


class RecordFinder
    constructor: -> super
    template: 'record-finder/field'
    templateModels: -> { field_name: @record_query.initial_field }

    events:
        "keyup .record-finder-query-string": "onKey"
        "click .record-finder-query": "displayFinder"

    session:
        include_associations: 'array'
        record_query: 'model'
        query_field: [ 'string', false, 'code' ]

    initialize:(options)->
        @record_query = new Lanes.Models.Query(fields: options.fields, collection_class: options.query_using)

    displayFinder: ->
        finder = new FinderDialog( title: @title, record_query: @record_query )

        finder.show().then( (dlg)->
            dlg.remove().record
        ).then( (record)=>
            record.withAssociations(@.includeAssociations...) if record
        ).then( (record)=>
            this.$el.trigger("display-record", record) if record
        , (e)->Lanes.warn(e) )

    onKey: (ev)->
        qf=this.$('.record-finder-query-string')
        if 13 == ev.keyCode
            this.runQuery(ev)
        else if qf.val().match( Lanes.Models.mixins.Lanes.sCodeField.INVALID )
            qf.val( qf.val().replace( Lanes.Models.mixins.Lanes.sCodeField.INVALID, '' ) )

    runQuery: (ev)->
        code = this.$(ev.target).val()
        @record_query.loadSingle(code.toUpperCase(),{ include: @includeAssociations })
            .then( (reply)=> this.$el.trigger("display-record", reply.record) if reply.record )

Lanes.Component.RecordFinder = Lanes.Component.Base.extend(RecordFinder)
