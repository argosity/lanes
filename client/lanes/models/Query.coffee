##= require_self
##= require ./query/Result
##= require ./query/ArrayResult
##= require ./query/CollectionResult

class Field extends Lanes.Models.Base
    registerForPubSub: false

    constructor: (attributes) ->
        super( _.defaults( attributes, {
            title: _.titleize(_.humanize(attributes.id))
        }))

    session:
        id:         type: 'string'
        title:      type: 'string'
        selected:   type: 'boolean'
        visible:    type: 'boolean', default: true
        query:      type: 'boolean', default: true
        editable:   type: 'boolean', default: true
        sortable:   type: 'boolean', default: true
        format:     type: 'function'
        flex:       type: 'number',  default: 1
        fixedWidth: type: 'number'
        textAlign:  type: 'string', default: 'left'
        sortBy:     type: 'any'
        component:  type: 'function'
        onColumnClick: type: 'function'

    derived:
        default_value: deps: ['id'], fn: ->
            switch @model_field.type
                when 'integer' then 0
                else ''
        model_field:
            deps: ['id'], fn: ->
                this.collection.query.model::_definition[@id]
        type:
            deps: ['model_field'], fn: ->
                type = @model_field?.type || 'string'
                if type == "code" then "string" else type
        fetchIndex:
            fn: ->
                fetchIndex = 0
                for field, index in this.collection.models
                    return fetchIndex if this == field
                    fetchIndex += 1 if field.query

    validValue: (value) ->
        if this.type == 'n'
            ! _.isNaN( parseFloat(value) )
        else
            value


class AvailableFields extends Lanes.Models.Collection

    model: Field

    constructor: (models, options) ->
        @query = options.query
        this.on('change', (changing) ->
            return unless changing.selected
            for model in @models
                model.selected = false unless model == changing
        )
        super
        @visible = @subcollection(where: {visible: true})


class Operator extends Lanes.Models.Base
    registerForPubSub: false

    session:
        id:       'string'
        name:     'string'
        types:    'array'
        field:    'object'
        selected: 'boolean'

    validForField: (field) ->
        _.isEmpty(@types) or _.includes(this.types, field.type)


class Operators extends Lanes.Models.Collection
    model: Operator

    constructor: (models, attrs) ->
        super
        this.add([
            { id: 'like', name: 'Starts With', types: Lanes.Models.Query.LIKE_QUERY_TYPES }
            { id: 'eq',   name: 'Equals' }
            { id: 'lt',   name: 'Less Than', types: Lanes.Models.Query.LESS_THAN_QUERY_TYPES }
            { id: 'gt',   name: 'More Than', types: Lanes.Models.Query.LESS_THAN_QUERY_TYPES }
        ])
        this.on('change', (changing) ->
            return unless changing.selected
            for model in @models
                model.selected = false unless model == changing
        )
        @setField(attrs.parent.query.initialField)

    setField: (field) ->
        @field = field
        @valid = @subcollection(filter: (op) => op.validForField(@field) )
        selected = this.findWhere(selected: true)
        unless selected && selected.validForField(field)
            @valid.at(0)?.selected = true


class Clause extends Lanes.Models.Base
    registerForPubSub: false

    session:
        value     : { type: 'string', default: '' }
        field     : 'state'
        operator  : 'state'

    associations:
        operators : { collection: Operators }
        fields:     { collection: AvailableFields }

    derived:
        description:
            deps: ['field', 'operator']
            fn: -> "#{@field.title} #{@operator?.id}"

        isValid:
            deps: ['field', 'operator', 'value']
            fn: -> this.field.validValue(@value)

        query:
            fn: -> @collection.query

    constructor: (options) ->
        super
        this.fields.reset(options.available_fields.models)
        @fields.on('change:selected',    this.setField,    this)
        @operators.on('change:selected', this.setOperator, this)
        @operator = @operators.findWhere(selected: true)
        if (field = @fields.findWhere(visible: true))
            field.selected = true
            @setField(field)

    setFromView: (type, val) ->
        if type == "fields" || type == "operators"
            this[type].get(val).selected = true
        else
            super

    setField: (field) ->
        return unless field.selected
        @operators.setField( field )
        @field = field

    setOperator: (operator) ->
        return unless operator.selected
        this.operator = operator

    remove: ->
        @collection.remove(this)

    toParam: ->
        param = {}
        op = this.operator.id
        value = this.get('value')
        value += '%' if 'like' == op
        value = parseFloat(value) if @field.type == "n"
        param[ @field.id ] = if 'eq' == op then value else { op: op, value: value }
        param


class Clauses extends Lanes.Models.Collection

    model: Clause

    initialize:(models, options) ->
        super
        @query = options.query
        @fields = options.query.fields

    session:
        field: 'state'

# needs to inherit from Base so network events will be listened to
class Lanes.Models.Query extends Lanes.Models.Base
    registerForPubSub: false

    @LIKE_QUERY_TYPES: ['string', 'code', 'visible_id']
    @LESS_THAN_QUERY_TYPES: ['integer', 'bigdec', 'number']
    @GREATER_THAN_QUERY_TYPES: ['integer', 'bigdec', 'number']

    @mergedSyncOptions: (args...) ->
        _.merge {}, args..., (a, b) ->
            return a.concat(b) if _.isArray(a)

    session:
        src:     'any'
        initialField: 'state'
        idIndex: 'number'
        initialFieldIndex: 'number'
        pageSize: { type: 'number', default: 100 }
        syncOptions: 'any'
        autoRetrieve: {type:'boolean', default: false, required: true}
        title: { type: 'string', default: 'Find Record' }
        sortField: 'object'
        defaultSort: 'any'
        changeCount: {type: 'integer', default: 0}
        sortAscending: ['boolean', true, true]

    associations:
        fields:  {collection: AvailableFields, inverse: 'query'}
        clauses: {collection: Clauses, inverse: 'query'}

    derived:
        isCollection:
            deps: ['src'], fn: -> Lanes.u.isCollection(@src)

        model:
            deps: ['src'], fn: ->
                return null unless @src
                if @isCollection then @src.model else @src

        idAttribute:
            deps: ['model'], fn: -> @model::idAttribute

        results:
            deps: ['src'], fn: ->
                if @isCollection
                    new Lanes.Models.Query.CollectionResult(this)
                else
                    new Lanes.Models.Query.ArrayResult(this, pageSize: @pageSize)

    events:
        'change:sortField':     '_updateSort'
        'change:sortAscending': '_updateSort'

    constructor: (options = {}) ->
        super

        @fields.reset()
        for col, i in options.fields
            rec = if _.isObject(col) then col else { id: col }
            @fields.add rec
            @idIndex = i if rec.id == @idAttribute
        unless @idIndex?
            @idIndex = @fields.length
            @fields.add(id: @idAttribute)
        # @clauses = new Clauses([], query: this )
        this.listenTo(@clauses, 'change remove reset', =>
            @results.reset()
            @results.ensureLoaded() if this.autoRetrieve
        )
        this.on('change:src change:results', ->
            this.trigger('load')
        )
        this.on('load', ->
            @changeCount += 1
        )

        if @initialFieldIndex
            @initialField = this.fields.visible.at(@initialFieldIndex)

        @initialField ||= this.fields.visible.findWhere(id: "code") ||
            this.fields.visible.findWhere(id: "visible_id") ||
            this.fields.visible.first()
        @reset(true)
        this

    clonedAttributes: ->
        attrs = @getAttributes(session: true, derived: false)
        attrs.fields = this.fields.serialize({session: true})
        attrs

    reset: (silent = false) ->
        unless @defaultSort is false
            sort = @defaultSort or @fields.findWhere(visible: true).id
            @setSortField( @fields.findWhere(id: sort), sortAscending: @sortAscending, silent: true )

        @clauses.reset([
            {query: this, available_fields: @fields, field: @initialField}
        ], {silent})

    setSortField: (field, options = {silent: false}) ->
        options.sortAscending ?= (if @sortField is field then !@sortAscending else true)
        @set({
            sortAscending: options.sortAscending
            sortField: field
        }, options)

    ensureLoaded: -> @results.ensureLoaded()

    isValid: ->
        ! @clauses.findWhere( isValid: false )

    loadModel: (model, options = {}) ->
        _.extend(options, _.result(this, 'syncOptions'), force: true)
        model.withAssociations(options.include || [], options)

    loadSingle: (code, options = {}) ->
        options.query = {}
        options.query[ @initialField.id ] = code
        _.extend(options, _.result(this, 'syncOptions'))
        @trigger('request')
        @src.fetch(options).then (model) =>
            @trigger('load')
            model

    defaultField: ->
        @fields.findWhere( field: @initialField )

    addNewClause: ->
        @clauses.add(query: this, available_fields: @fields, field: @initialField)

    _updateSort: _.debounce ->
        @results._updateSort()
