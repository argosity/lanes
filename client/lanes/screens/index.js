import { observable, autorun } from 'mobx';
import { map, uniq } from 'lodash';
import Group from './group';
import Config from '../config';
import Sync from '../models/sync';

import user from '../user';

const Screens = observable({

    all: observable.map(),

    active: observable.array(),

    groups: observable.map(),

    get activeGroups() {
        return uniq(map(this.active, s => Group.forId(s.group_id)));
    },

    refresh() {
        Sync.perform(`${Config.api_path}/lanes/screens`).then(({ data: { screens: screenIds } }) => {
            this.active.replace(map(screenIds, id => this.all.get(id)));
        });
    },

});
export default Screens;

let previousLoggedIn = null;
autorun(() => {
    if (user.isLoggedIn !== previousLoggedIn){
        previousLoggedIn = user.isLoggedIn;
        Screens.refresh();
    }
});
