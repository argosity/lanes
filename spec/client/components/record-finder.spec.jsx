import React from 'react';
import { range } from 'lodash';


import Query     from 'lanes/models/query';

import { Snapshot } from 'lanes/testing/screens';
import RecordFinder from 'lanes/components/record-finder';

import { Container } from '../test-models';

jest.mock('lanes/models/sync');

describe("RecordFinder Component", () => {
    let query;
    let props;
    beforeEach(() => {
        query = new Query({
            src: Container,
            fields: [
                { id: 'id', visible: false, queryable: false },
                { id: 'computed', title: 'White?', loadable: false },
                { id: 'name' },
                { id: 'location' },
            ],
        });
        props = {
            onRecordFound: jest.fn(),
            name: 'name',
            fields: {
                name: {
                    value: 'ONE',
                    events: {}
                }
            },
            query
        }

        range(0, 5).forEach(
            i => query.results.rows.push([i, `name ${i}`, `location ${i}`]),
        );
    });

    it('renders', () => {
        const finder = mount(
            <RecordFinder {...props} />,
        );
        console.log(finder.debug())
        expect(finder).toHaveRendered('TextInput');
    });
});
