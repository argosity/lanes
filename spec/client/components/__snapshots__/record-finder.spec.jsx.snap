// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`RecordFinder Component renders 1`] = `
<div
  className="form-field text record-finder"
>
  <div
    className="grommetux-form-field grommetux-form-field--size-medium"
    onClick={[Function]}
  >
    <label
      className="grommetux-form-field__label"
      htmlFor={undefined}
    >
      Name
    </label>
    <span
      className="grommetux-form-field__contents"
    >
      <input
        autoComplete="off"
        autoFocus={undefined}
        className="grommetux-text-input grommetux-input"
        defaultValue={undefined}
        name="name"
        onBlur={[Function]}
        onChange={[Function]}
        onFocus={[Function]}
        onKeyDown={[Function]}
        onKeyPress={[Function]}
        placeholder={undefined}
        readOnly={false}
        tabIndex={0}
        type="text"
        value=""
      />
      <button
        aria-label={undefined}
        className="grommetux-button grommetux-button--plain grommetux-control-icon-search"
        disabled={false}
        href={undefined}
        onBlur={[Function]}
        onClick={[Function]}
        onFocus={[Function]}
        onMouseDown={[Function]}
        onMouseUp={[Function]}
        type="button"
      >
        <span
          className="grommetux-button__icon"
        >
          <svg
            aria-label="search"
            className="grommetux-control-icon grommetux-control-icon-search grommetux-control-icon--responsive"
            height="24px"
            role="img"
            version="1.1"
            viewBox="0 0 24 24"
            width="24px"
          >
            <path
              d="M15,15 L22,22 L15,15 Z M9.5,17 C13.6421356,17 17,13.6421356 17,9.5 C17,5.35786438 13.6421356,2 9.5,2 C5.35786438,2 2,5.35786438 2,9.5 C2,13.6421356 5.35786438,17 9.5,17 Z"
              fill="none"
              stroke="#000"
              strokeWidth="2"
            />
          </svg>
        </span>
      </button>
    </span>
  </div>
</div>
`;
